README.md
Traffic Forecast
===========
![PHP status](https://github.com/Trailblazers/Traffic-Forecast/workflows/PHP/badge.svg)
![ESLint status](https://github.com/Trailblazers/Traffic-Forecast/workflows/JS/badge.svg)
![Active installs](https://img.shields.io/wordpress/plugin/installs/Traffic-Forecast.svg)
![Downloads](https://img.shields.io/wordpress/plugin/dt/Traffic-Forecast.svg)
[![Rating](https://img.shields.io/wordpress/plugin/r/Traffic-Forecast.svg)](https://wordpress.org/support/plugin/Traffic-Forecast/reviews/)
[![License: GPLv3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

[Traffic Forecast](https://www.TrafficForecast.com/) is an open-source and privacy-friendly Forecast plugin for WordPress. 

![Screenshot of the Traffic Forecast dashboard](https://github.com/Trailblazers/Traffic-Forecast/raw/master/assets/src/img/screenshot-1.png?v=1)

## Features

- No external services. Data is yours and yours alone.
- No personal information or anything visitor specific is tracked.
- Fast! Handles thousands of daily visitors or sudden bursts of traffic without breaking a sweat.
- Lightweight. Adds only 985 bytes of data to your pages.
- Plug and play. Just install and activate the plugin and stats will be recorded right away.
- Open-source (GPLv3 licensed).

### How to install

To run the latest development version of the plugin, take the following steps.

First, clone the repository using Git in your `/wp-content/plugins/` directory
```
git clone git@github.com:Trailblazers/Traffic-Forecast.git
```

Create the autoloader using Composer.
```
composer install
```

Install client-side dependencies using NPM
```
npm install
```

Build the plugin assets by issuing the following command:
``` 
npm run build
```

### Usage

Stats will be collected right away after you install and activate the plugin. You can view your stats on the **Dashboard > Forecast** page.

### Contributing

You can contribute to Traffic Forecast in many different ways. For example:


- Write about the plugin on your blog or share it on social media.

- Help fund development and support costs through [Traffic Forecast on OpenCollective](https://opencollective.com/Traffic-Forecast).

### License

GNU General Public License v3.0


CHANGELOG.md
# Changelog

### 0.0.1 - Apr 25, 2023

- Fix issue with dashboard widget rendering incorrectly when it was initially hidden.
- Only render chart in dashboard widget when it is currently visible.


### 1.0.37 - Dec 07, 2022

- Fix potential issue in tracking script causing incorrect stats collection.


### 1.0.36 - Dec 06, 2022

- Shrink size of tracking script to 985 bytes (when using gzip compression).


### 1.0.35 - Oct 21, 2022

- You can now call `Traffic_Forecast.trackPageview( postId )` to manually track a pageview. This is especially useful for Single Page Applications.
- Update referrer blocklist.
- If referrer URL is on referrer blocklist, the entire request is now ignored (vs. just the referrer part).
- Update JavaScript dependencies.


#### 1.0.34 - Sep 19, 2022

- Track visits to front page (if not a static page) as well.
- Use gmdate() for determining whether to send cache headers.
- Update JavaScript dependencies.


#### 1.0.33 - Aug 17, 2022

- Always show at least 1 visitor if pageviews were recorded that day.
- Decrease Cache-Control header of REST API responses to 60s.
- Add button to reset all statistics.
- Add filter `Traffic_Forecast_url_aggregations` to modify or add URL aggregations for referrer URL's.
- Created an initiative on OpenCollective for others to help fund development and support costs of Traffic Forecast. Please visit https://opencollective.com/Traffic-Forecast for more details.


#### 1.0.32 - Jun 2, 2022

- Fix error in latest release by ensuring Endpoint_Installer class is always loaded when needed.


#### 1.0.30 - Jun 1, 2022

- Delete option `Traffic_Forecast_use_custom_endpoint` on plugin uninstall.
- Only load dashboard-widget.js if user has required capability for viewing stats.
- Add constant `Traffic_Forecast_CUSTOM_ENDPOINT` for  [disabling the custom endpoint](https://github.com/Trailblazers/Traffic-Forecast/blob/master/code-snippets/disable-custom-endpoint.php) entirely. This filter can also be used to [manually install the endpoint file to a different file location](https://github.com/Trailblazers/Traffic-Forecast/blob/master/code-snippets/use-different-custom-endpoint.php).
- Re-attempt installation of the custom endpoint file every hour. This also automatically re-tests whether the custom endpoint is still working as expected. The plugin already did this whenever you visited the dashboard page, but now it simply runs more often and without requiring you to log-in to your WordPress admin.


#### 1.0.29 - Jan 25, 2022

- Skip empty lines in buffer file to avoid a PHP notice in certain edge cases.
- Make use of JS library for date presets.
- Filter out user agents containing "preview".
- Normalize referrer URl's for Yahoo search results.
- Add class attribute to cron warning so it can be hidden using CSS by targeting `.Traffic-Forecast-cron-warning`.


#### 1.0.28 - Sep 27, 2021

- Account for missing configuration object, for example if theme does not properly call `wp_head()`.
- Cast database result to integer types so we do not have to do it client-side.
- Cache dashboard requests to REST API for 5 minutes (using browser cache).
- Use integers for viewbox coordinates (for increased performance).
- Simple locking mechanism for aggregation job to detect previous runs which are still busy.
- Add WP CLI command for aggregating stats without having to go through WP Cron: `wp Traffic-Forecast aggregate` 


#### 1.0.27 - May 4, 2021

- Use `defer` attribute on script to not block parsing at all.
- Normalize Pinterest URL's with and without www subdomain.
- Fix some missing translations.


#### 1.0.26 - Feb 17, 2021

- Re-introduce last 28 days by popular request.
- Fix date presets dropdown not working in WebKit based browsers.
- Update built-in referrer blocklist.


#### 1.0.25 - Feb 16, 2021

- Add several new date presets, like yesterday or last week.
- Add support for entering custom date ranges using keyboard input.
- Improve algorithm for chart y-axes to prevent huge jumps.
- Add function for manually printing tracking script: `<?php Traffic_analyics_tracking_script(); ?>`
- Update JS dependencies.


#### 1.0.24 - Jan 20, 2021

- Append random query parameter to tracking request to avoid beind cached now that the plugin is using a custom tracking endpoint.
- Update JS dependencies.


#### 1.0.23 - Jan 15, 2021

Major performance improvement by writing an optimized endpoint file containing the correct file paths, regardless of WordPress directory structure set-up. 

If your WordPress root directory is not writable, go to the Traffic Forecast settings page for instructions on how to manually create this optimized endpoint file.


#### 1.0.22 - Nov 18, 2020

- Fix PHP warnings in migration script, for trying to unlink unexisting file  (eg on new plugin installs).
- Generate URL to custom endpoint file correctly. Fixes issue with some multilingual plugins not recording pageviews correctly.


#### 1.0.21 - Nov 5, 2020

- Update option that stores whether to use custom endpoint after unlinking it, to prevent sending requests to an unexisting file.


#### 1.0.20 - Nov 3, 20202

- Do not use custom endpoint file if using custom uploads directory.
- Use value from `Traffic_Forecast_USE_CUSTOM_ENDPOINT` if it is defined.
- Only call add_cap on administrator role if such a role exists 
- Update JS dependencies.
- Update built-in referrer blocklist.


#### 1.0.19 - Sep 2, 2020

- Create buffer file directory if it does not exist yet, eg on a fresh WP install. 
- Update preact and date-fns to their latest versions.
- Update built-in referrer blocklist.


#### 1.0.18 - Aug 25, 2020

- Fix issue with tracking not working on AMP powered pages or issuing a request to a non-existing file on cdn.ampproject.org.


#### 1.0.17 - Aug 19, 2020

- Fix issue when using Modern color scheme introduced in WordPress 5.5.
- Improve test for custom endpoint file by checking for exact response body.

src/class-admin.php
<?php
/**
 * @package Traffic-Forecast
 * @license GPL-3.0+
 * @author TrailBlazers
 */
namespace TrafficForecast;

class Admin
{
	public function init()
	{
		global $pagenow;

		add_action( 'init', array( $this, 'maybe_run_migrations' ) );
		add_action( 'admin_menu', array( $this, 'register_menu' ) );
		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
		add_action( 'wp_dashboard_setup', array( $this, 'register_dashboard_widget' ) );

		switch ( $pagenow ) {
			case 'index.php':
				// Hooks for main dashboard page
				add_action( 'shutdown', array( $this, 'maybe_run_endpoint_installer' ) );
				add_action( 'init', array( $this, 'maybe_seed' ) );
				break;

			case 'plugins.php':
				// Hooks for plugins overview page
				add_filter( 'plugin_action_links_' . plugin_basename( Traffic_Forecast_PLUGIN_FILE ), array( $this, 'add_plugin_settings_link' ), 10, 2 );
				add_filter( 'plugin_row_meta', array( $this, 'add_plugin_meta_links' ), 10, 2 );
				break;
		}

	}

	public function register_menu()
	{
		add_submenu_page( 'index.php', esc_html__( 'Traffic Forecast', 'Traffic-Forecast' ), esc_html__( 'Forecast', 'Traffic-Forecast' ), 'view_Traffic_Forecast', 'Traffic-Forecast', array( $this, 'show_page' ) );
	}

	public function enqueue_scripts( $suffix )
	{
		// do not load any scripts if user is missing required capability for viewing
		if ( ! current_user_can( 'view_Traffic_Forecast' ) ) {
			return;
		}

		switch ( $suffix ) {
			case 'index.php':
				// load scripts for dashboard widget
				wp_enqueue_script( 'Traffic-Forecast-dashboard-widget', plugins_url( '/assets/dist/js/dashboard-widget.js', Traffic_Forecast_PLUGIN_FILE ), array( 'wp-i18n' ), Traffic_Forecast_VERSION, true );
				if ( function_exists( 'wp_set_script_translations' ) ) {
					wp_set_script_translations( 'Traffic-Forecast-dashboard-widget', 'Traffic-Forecast' );
				}
				wp_localize_script(
					'Traffic-Forecast-dashboard-widget',
					'Traffic_Forecast',
					array(
						'root' => rest_url(),
						'nonce' => wp_create_nonce( 'wp_rest' ),
						'colors' => $this->get_colors(),
					)
				);
				break;

			case 'dashboard_page_Traffic-Forecast':
				$user_roles = $this->get_available_roles();
				$start_of_week = (int) get_option( 'start_of_week' );
				$settings = get_settings();
				$colors = $this->get_colors();
				$endpoint_installer = new Endpoint_Installer();

				wp_enqueue_script( 'Traffic-Forecast-admin', plugins_url( 'assets/dist/js/admin.js', Traffic_Forecast_PLUGIN_FILE ), array( 'wp-i18n' ), Traffic_Forecast_VERSION, true );
				if ( function_exists( 'wp_set_script_translations' ) ) {
					wp_set_script_translations( 'Traffic-Forecast-admin', 'Traffic-Forecast' );
				}
				wp_localize_script('Traffic-Forecast-admin', 'Traffic_Forecast', array(
					'root'          => rest_url(),
					'nonce'         => wp_create_nonce( 'wp_rest' ),
					'start_of_week' => (int) $start_of_week,
					'user_roles'    => $user_roles,
					'settings'      => $settings,
					'showSettings'  => current_user_can( 'manage_Traffic_Forecast' ),
					'dbSize' => $this->get_database_size(),
					'colors' => $colors,
					'multisite' => is_multisite(),
					'custom_endpoint' => array(
						'enabled' => using_custom_endpoint(),
						'file_contents' => $endpoint_installer->get_file_contents(),
						'wp_root_dir' => rtrim( ABSPATH, '/' ),
					),
				));
				break;
		}
	}

	private function get_available_roles()
	{
		$roles = array();
		foreach ( wp_roles()->roles as $key => $role ) {
			$roles[ $key ] = $role['name'];
		}
		return $roles;
	}

	private function is_cron_event_working()
	{
		// detect issues with WP Cron event not running
		// it should run every minute, so if it didn't run in 10 minutes there is most likely something wrong
		$next_scheduled = wp_next_scheduled( 'Traffic_Forecast_aggregate_stats' );
		return $next_scheduled !== false && $next_scheduled > ( time() - HOUR_IN_SECONDS );
	}

	public function show_page()
	{
		// aggregate stats whenever this page is requested
		do_action( 'Traffic_Forecast_aggregate_stats' );

		// determine whether buffer file is writable
		$buffer_filename = get_buffer_filename();
		$buffer_dirname = dirname( $buffer_filename );
		$is_buffer_dir_writable = wp_mkdir_p( $buffer_dirname ) && is_writable( $buffer_dirname );

		// determine whether cron event is set-up properly seeand running in-time
		$is_cron_event_working = $this->is_cron_event_working();

		require Traffic_Forecast_PLUGIN_DIR . '/views/admin-page.php';
		add_action( 'admin_footer_text', array( $this, 'footer_text' ) );
	}

	public function footer_text()
	{
		/* translators: %1$s links to the WordPress.org plugin review page, %2$s links to the admin page for creating a new post */
		return sprintf( wp_kses( __( 'If you enjoy using Traffic Forecast, please <a href="%1$s">review the plugin on WordPress.org</a> or <a href="%2$s">write about it on your blog</a> to help out.', 'Traffic-Forecast' ), array( 'a' => array( 'href' => array() ) ) ), 'https://wordpress.org/support/view/plugin-reviews/Traffic-Forecast?rate=5#postform', admin_url( 'post-new.php' ) );
	}

	public function maybe_run_endpoint_installer()
	{
		if ( ! isset( $_GET['page'] ) || $_GET['page'] !== 'Traffic-Forecast' ) {
			return;
		}

		// do not run if Traffic_Forecast_CUSTOM_ENDPOINT is defined
		if ( defined( 'Traffic_Forecast_CUSTOM_ENDPOINT' ) ) {
			return;
		}

		install_and_test_custom_endpoint();
	}

	public function maybe_run_migrations()
	{
		if ( ! current_user_can( 'manage_options' ) ) {
			return;
		}

		$from_version = isset( $_GET['Traffic_Forecast_migrate_from_version'] ) ? $_GET['Traffic_Forecast_migrate_from_version'] : get_option( 'Traffic_Forecast_version', '0.0.1' );
		$to_version = Traffic_Forecast_VERSION;
		if ( version_compare( $from_version, $to_version, '>=' ) ) {
			return;
		}

		// run upgrade migrations (if any)
		$migrations_dir = Traffic_Forecast_PLUGIN_DIR . '/migrations/';
		$migrations = new Migrations( $from_version, $to_version, $migrations_dir );
		$migrations->run();
		update_option( 'Traffic_Forecast_version', $to_version );

		// make sure scheduled event is set-up correctly
		$aggregator = new Aggregator();
		$aggregator->setup_scheduled_event();
	}

	private function get_colors()
	{
		$color_scheme_name = get_user_option( 'admin_color' );
		global $_wp_admin_css_colors;
		if ( empty( $_wp_admin_css_colors[ $color_scheme_name ] ) ) {
			$color_scheme_name = 'fresh';
		}

		return $_wp_admin_css_colors[ $color_scheme_name ]->colors;
	}

	public function register_dashboard_widget()
	{
		// only show if user can view stats
		if ( ! current_user_can( 'view_Traffic_Forecast' ) ) {
			return;
		}

		add_meta_box( 'Traffic-Forecast-dashboard-widget', 'Traffic Forecast', array( $this, 'dashboard_widget' ), 'dashboard', 'side', 'high' );
	}

	public function dashboard_widget()
	{
		echo '<div id="Traffic-Forecast-dashboard-widget-mount"></div>';
		echo sprintf( '<p class="help" style="text-align: center;">%s &mdash; <a href="%s">%s</a></p>', esc_html__( 'Showing site visits over last 14 days', 'Traffic-Forecast' ), esc_attr( admin_url( 'index.php?page=Traffic-Forecast' ) ), esc_html__( 'View all statistics', 'Traffic-Forecast' ) );
	}

	/**
	 * Add the settings link to the Plugins overview
	 *
	 * @param array $links
	 * @param       $file
	 *
	 * @return array
	 */
	public function add_plugin_settings_link( $links, $file )
	{
		$settings_link = sprintf( '<a href="%s">%s</a>', admin_url( 'index.php?page=Traffic-Forecast#/settings' ), esc_html__( 'Settings', 'Traffic-Forecast' ) );
		array_unshift( $links, $settings_link );
		return $links;
	}

	/**
	 * Adds meta links to the plugin in the WP Admin > Plugins screen
	 *
	 * @param array $links
	 * @param string $file
	 *
	 * @return array
	 */
	public function add_plugin_meta_links( $links, $file )
	{
		if ( $file !== plugin_basename( Traffic_Forecast_PLUGIN_FILE ) ) {
			return $links;
		}

		$links[] = '<a href="https://www.TrafficForecast.com/">' . esc_html__( 'Visit plugin site', 'Traffic-Forecast' ) . '</a>';
		return $links;
	}

	public function get_database_size()
	{
		global $wpdb;
		$sql = $wpdb->prepare(
			'
			SELECT ROUND(SUM((DATA_LENGTH + INDEX_LENGTH) / 1024 / 1024), 2)
			FROM information_schema.TABLES
			WHERE TABLE_SCHEMA = %s AND TABLE_NAME LIKE %s',
			DB_NAME,
			$wpdb->prefix . 'Traffic_Forecast_%'
		);

		return $wpdb->get_var( $sql );
	}

	public function maybe_seed()
	{
		global $wpdb;

		if ( ! isset( $_GET['Traffic_Forecast_seed'] ) || ! current_user_can( 'manage_options' ) ) {
			return;
		}

		$query = new \WP_Query();
		$posts = $query->query(
			array(
				'posts_per_page' => 32,
				'post_type' => 'any',
				'post_status' => 'publish',
			)
		);
		$post_count = count( $posts );
		$referrer_urls = array();
		$sample_referrers = array(
			'https://www.wordpress.org',
			'https://www.wordpress.org/plugins/Traffic-Forecast',
			'https://www.Trailblazers.com',
			'https://duckduckgo.com',
			'https://www.mozilla.org',
			'https://www.eff.org',
			'https://letsencrypt.org',
			'https://TrailBlazers.com',
			'https://github.com/Trailblazers/Traffic-Forecast',
			'https://lobste.rs',
			'https://joinmastodon.org',
			'https://www.php.net',
			'https://mariadb.org',
			'https://referrer-1.com',
			'https://referrer-2.com',
			'https://referrer-3.com',
			'https://referrer-4.com',
			'https://referrer-5.com',
			'https://referrer-6.com',
			'https://referrer-7.com',
			'https://referrer-8.com',
			'https://referrer-9.com',
			'https://referrer-10.com',
			'https://referrer-11.com',
			'https://referrer-12.com',
			'https://referrer-13.com',
			'https://referrer-14.com',
			'https://referrer-15.com',
			'https://referrer-16.com',
			'https://referrer-17.com',
			'https://referrer-18.com',
			'https://referrer-19.com',
			'https://referrer-20.com',
			'https://t.co/IiADWZC13f',
			'https://www.reddit.com/r/Wordpress/comments/e6ycsm/privacy_friendly_Forecast_plugin_that_does_not/',
			'android-app://com.stefandekanski.hackernews.free',
		);
		foreach ( $sample_referrers as $url ) {
			$wpdb->insert(
				$wpdb->prefix . 'Traffic_Forecast_referrer_urls',
				array(
					'url' => $url,
				)
			);
			$referrer_urls[ $wpdb->insert_id ] = $url;
		}
		$referrer_count = count( $referrer_urls );

		$n = 3 * 365;
		for ( $i = 0; $i < $n; $i++ ) {
			$progress = ( $n - $i ) / $n;
			$date = gmdate( 'Y-m-d', strtotime( sprintf( '-%d days', $i ) ) );
			$pageviews = max( 1, rand( 500, 1000 ) * $progress ^ 2 );
			$visitors = max( 1, $pageviews * rand( 3, 6 ) / 10 );

			// simulate a huge peak in traffic every 180 days
			if ( rand( 1, 180 ) === 1 ) {
				$pageviews = $pageviews * 10;
				$visitors = $visitors * 10;
			}

			$wpdb->insert(
				$wpdb->prefix . 'Traffic_Forecast_site_stats',
				array(
					'date' => $date,
					'pageviews' => $pageviews,
					'visitors' => $visitors,
				)
			);

			$values = array();
			foreach ( $posts as $post ) {
				array_push( $values, $date, $post->ID, round( $pageviews / $post_count * rand( 5, 15 ) / 10 ), round( $visitors / $post_count * rand( 5, 15 ) / 10 ) );
			}
			$placeholders = rtrim( str_repeat( '(%s,%d,%d,%d),', count( $posts ) ), ',' );
			$wpdb->query( $wpdb->prepare( "INSERT INTO {$wpdb->prefix}Traffic_Forecast_post_stats(date, id, pageviews, visitors) VALUES {$placeholders}", $values ) );

			$values = array();
			foreach ( $referrer_urls as $id => $referrer_url ) {
				array_push( $values, $date, $id, round( $pageviews / $referrer_count * rand( 5, 15 ) / 10 ), round( $visitors / $referrer_count * rand( 5, 15 ) / 10 ) );
			}
			$placeholders = rtrim( str_repeat( '(%s,%d,%d,%d),', count( $referrer_urls ) ), ',' );
			$wpdb->query( $wpdb->prepare( "INSERT INTO {$wpdb->prefix}Traffic_Forecast_referrer_stats(date, id, pageviews, visitors) VALUES {$placeholders}", $values ) );
		}
	}
}
